installer: Don't run as root

worldofpeace opened the pull request:
###### Motivation for this change

There's many reason why it is and is going to
continue to be difficult to do this:

1. All display-managers (excluding slim) default PAM rules
   disallow root auto login.

2. We can't use wayland

3. We have to use system-wide pulseaudio

4. It could break applications in the session.
   This happened to dolphin in plasma5
   in the past.

This is a growing technical debt, let's just use
passwordless sudo.

---

This is a per-requisite to having a GNOME3 iso.
Broken up from #66313

I've supplied all the requested changes to what was raised on this commit.
In particular  https://github.com/NixOS/nixpkgs/pull/66313#pullrequestreview-272620860.


###### Things done

I've built `iso_minimal` and it auto logs in as `live` when tested
in qemu.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

worldofpeace committed:
installer: Don't run as root

There's many reason why it is and is going to
continue to be difficult to do this:

1. All display-managers (excluding slim) default PAM rules
   disallow root auto login.

2. We can't use wayland

3. We have to use system-wide pulseaudio

4. It could break applications in the session.
   This happened to dolphin in plasma5
   in the past.

This is a growing technical debt, let's just use
passwordless sudo.

grahamc commented on nixos/modules/installer/cd-dvd/installation-cd-graphical-base.nix:
```diff
@@ -8,6 +8,16 @@ with lib;
   imports = [ ./installation-cd-base.nix ];
 
+  # Whitelist wheel users to do anything
+  # This is useful for things like pkexec
```
Maybe should add a caution here that this is explicitly dangerous, and should only be used for livecds. ... Just in case it gets copy-pasted somewhere.

grahamc commented on nixos/modules/installer/cd-dvd/installation-cd-graphical-base.nix:
```diff
@@ -17,7 +27,7 @@ with lib;
     # Automatically login as root.
     displayManager.slim = {
```
I won't ask it of this PR, but slim is long abandoned. We  should get off of it sometime :)

worldofpeace replied:
Yep, I raised this in https://github.com/NixOS/nixpkgs/pull/66313#issue-305402121 at the very end. Some people had strong opinions though in the past IIRC.

grahamc reviewed (approved):
Shall we go for slightly more branding and call the "live" user "nixos"?

worldofpeace commented:
> Shall we go for slightly more branding and call the "live" user "nixos"?

When adisbladis and I discussed this and he authored the original change
I think @davidak suggested to use `live` https://github.com/NixOS/nixpkgs/pull/42610#issuecomment-404222033.

grahamc commented:
It sounds like the name then was something with a hyphen, but I can't see what it was. I still like `nixos` as the name, but :man_shrugging:  :wink: .

One thing I spotted in that linked PR is the suggestion of `sudo -i`, which we probably should document.

worldofpeace commented:
> It sounds like the name then was something with a hyphen, but I can't see what it was. I still like `nixos` as the name, but man_shrugging wink .
> 
> One thing I spotted in that linked PR is the suggestion of `sudo -i`, which we probably should document.

I think I'll do this :+1:

worldofpeace commented on nixos/modules/installer/cd-dvd/installation-cd-graphical-kde.nix:
```diff
@@ -30,15 +30,27 @@ with lib;
       Version=1.0
       Type=Application
       Name=NixOS Manual
-      Exec=firefox ${config.system.build.manual.manualHTMLIndex}
```
Just realized that `nixos-manual.desktop` generation isn't needed anymore,
as `nixos-manual` has a desktop item that will intelligently launch the default browser.

I think I'll fix this in the second pr, dropping this bit.

(not the gparted part though)

lheckemann replied:
So this can be removed?

samueldr commented:
*(Noting for myself, mainly)* this will affect sd_image, meaning both that it needs to be tested, and (external) documentation amended as needed.

worldofpeace committed:
gparted: correct polkit support

Use wrapGAppsHook as well

worldofpeace commented:
https://github.com/NixOS/nixpkgs/pull/66338/commits/bd59a8b787e2e7cc78db36d082df4d5817d909be was needed so users can use gparted and it will call into `pkexec` so we don't have to use a custom `sudo` desktop item.

They'll need no kind of authentication because of the polkit rule we added.

bachp commented:
I remember that I once added the possibility do do root login into the install media via SSH. If the default user is no longer `root` than this is probably no longer needed.

The SSH install workflow would then just be as follows:

1. Set a password (or public key) for user `nixos`
2. Enable SSH daemon
3. Login via `ssh nixos@<target>`
4. Run `sudo -i`

So only password login for normal users is required.

worldofpeace commented:
Guess that means you authored

```
 <para>
    If you would like to continue the installation from a different machine you
    need to activate the SSH daemon via <literal>systemctl start
    sshd</literal>. In order to be able to login you also need to set a
    password for <literal>root</literal> using <literal>passwd</literal>.
   </para>

```

I guess we should document it this way, but I believe we should still permit root login with ssh.

worldofpeace commented:
I've now tested the graphical iso.

#### Things done

1. can `sudo -i` on tty
2. can run `sudo systemctl start display-manager` on tty
3. graphical session starts
4. all desktop items launched properly (gparted without password prompt from polkit)
5. can use `sudo` without password in `konsole`

Did the procedure to login via ssh for `root` and `nixos`.

grahamc commented on nixos/doc/manual/installation/installing.xml:
```diff
@@ -29,13 +29,14 @@
   <para>
-   You are logged-in automatically as <literal>root</literal>. (The
-   <literal>root</literal> user account has an empty password.)
+   You are logged-in automatically as <literal>nixos</literal>.
```
```suggestion
   You are logged-in automatically as the <literal>nixos</literal> user.
```

grahamc commented on nixos/doc/manual/installation/installing.xml:
```diff
@@ -65,9 +66,9 @@
-    sshd</literal>. In order to be able to login you also need to set a
-    password for <literal>root</literal> using <literal>passwd</literal>.
+    need to activate the SSH daemon via <command>systemctl start
+    sshd</command>. You then must set a password for either <literal>root</literal> or
```
What do you think about simplifying it and dropping the "or root" bit? It'll be simpler and clearer for the user, who won't have to choose which is better.

worldofpeace replied:
Well I'm thinking, someone tries to login as root and discovers "well I guess I have to set a password for root to if I want to use it?"

grahamc commented on nixos/doc/manual/installation/installing.xml:
```diff
@@ -65,9 +66,9 @@
-    password for <literal>root</literal> using <literal>passwd</literal>.
+    need to activate the SSH daemon via <command>systemctl start
+    sshd</command>. You then must set a password for either <literal>root</literal> or
+    <literal>nixos</literal> with <command>passwd></command> in order to login.
```
```suggestion
    <literal>nixos</literal> with <command>passwd></command> before you are able to login.
```

grahamc commented on nixos/doc/manual/release-notes/rl-1909.xml:
```diff
@@ -33,6 +33,12 @@
    </listitem>
+   <listitem>
+    <para>
+     The installer now uses a less privileged <literal>nixos</literal> user whereas before we logged in as root.
```
```suggestion
     The installer now uses a less privileged <literal>nixos</literal> user instead of <literal>root</literal>.
```

grahamc replied:
I drop "less privileged" because they're not actually less privileged, and saying it is might lead to questions about well how do I do the thing then.

grahamc reviewed (approved):
I haven't tested it, but I am quite sure it has been tested :) Looks great.

worldofpeace committed:
gparted: add adwaita-icon-theme

This adds the icon theme to XDG_DATA_DIRS.
It doesn't appear Plasma5 is properly configured for gtk
apps so this works around there being no icon theme installed
for it.

samueldr commented:
Just validated that sd_image works as expected.

worldofpeace commented:
Thanks everyone.

flokli commented:
There was some documentation fixes missing in `nixos/doc/manual/configuration/profiles/installation-device.xml` - I included them in https://github.com/NixOS/nixpkgs/pull/63773 now.
